sdk/core/core-client/src/deserializationPolicy.ts

```diff
@@ -118,24 +123,110 @@ async function deserializeResponseBody(
-    ? 200 <= parsedResponse.status && parsedResponse.status < 300
-    : !!responseSpec;
+  const errorResponse = handleErrorResponse(parsedResponse, operationSpec);
+  if (errorResponse) {
+    // There is no error spec. So, return it
+    return errorResponse;
+  }
 
   // There is no operation response spec for current status code.
   // So, treat it as an error case and use the default response spec to deserialize the response.
-  if (!isExpectedStatusCode) {
-    const defaultResponseSpec = operationSpec.responses.default;
-    if (!defaultResponseSpec) {
+
+  // An operation response spec does exist for current status code, so
+  // use it to deserialize the response.
+  if (responseSpec) {
+    if (responseSpec.bodyMapper) {
+      let valueToDeserialize: any = parsedResponse.parsedBody;
+      if (operationSpec.isXML && responseSpec.bodyMapper.type.name === MapperTypeNames.Sequence) {
+        valueToDeserialize =
+          typeof valueToDeserialize === "object"
+            ? valueToDeserialize[responseSpec.bodyMapper.xmlElementName!]
+            : [];
+      }
+      try {
+        parsedResponse.parsedBody = operationSpec.serializer.deserialize(
+          responseSpec.bodyMapper,
+          valueToDeserialize,
+          "operationRes.parsedBody"
+        );
+      } catch (error) {
+        const restError = new RestError(
+          `Error ${error} occurred in deserializing the responseBody - ${parsedResponse.bodyAsText}`,
+          {
+            statusCode: parsedResponse.status,
+            request: parsedResponse.request,
+            response: parsedResponse
+          }
+        );
+        throw restError;
+      }
+    } else if (operationSpec.httpMethod === "HEAD") {
+      // head methods never have a body, but we return a boolean to indicate presence/absence of the resource
+      parsedResponse.parsedBody = response.status >= 200 && response.status < 300;
+    }
+
+    if (responseSpec.headersMapper) {
+      parsedResponse.parsedHeaders = operationSpec.serializer.deserialize(
+        responseSpec.headersMapper,
+        parsedResponse.headers.toJSON(),
+        "operationRes.parsedHeaders"
+      );
+    }
+  }
+
+  return parsedResponse;
+}
+
+function handleErrorResponse(
+  parsedResponse: FullOperationResponse,
+  operationSpec: OperationSpec
+): FullOperationResponse | undefined {
+  enum RESPONSE_STATUS {
+    NONERROR,
+    ERRORNONDEFAULT,
+    ERRORDEFAULT
+  }
+  let parsedResponseStatus: RESPONSE_STATUS;
+  const statusCodes: string[] = Object.keys(operationSpec.responses);
+  const errorResponseCodes: string[] = [],
+    nonErrorResponseCodes: string[] = [];
+
+  for (const statusCode of statusCodes) {
+    if (statusCode !== "default" && operationSpec.responses[statusCode].isError) {
+      errorResponseCodes.push(statusCode);
+    } else {
+      nonErrorResponseCodes.push(statusCode);
+    }
+  }
+
+  if (
+    nonErrorResponseCodes.includes(parsedResponse.status + "") ||
```

Comment:
is `parsedResponse.status + ""` to convert number to string?  `toString()` for better readability?

Reply:
Changed the logic. So, this line no longer exists.